Tests/NIOPosixTests/IPv4Header.swift

```diff
@@ -0,0 +1,316 @@
+    var headerChecksum: UInt16
+    var sourceIpAddress: IPv4Address
+    var destinationIpAddress: IPv4Address
+    
+    fileprivate init(
+        versionAndIhl: UInt8,
+        dscpAndEcn: UInt8,
+        totalLength: UInt16,
+        identification: UInt16,
+        flagsAndFragmentOffset: UInt16,
+        timeToLive: UInt8,
+        `protocol`: NIOIPProtocol,
+        headerChecksum: UInt16,
+        sourceIpAddress: IPv4Address,
+        destinationIpAddress: IPv4Address
+    ) {
+        self.versionAndIhl = versionAndIhl
+        self.dscpAndEcn = dscpAndEcn
+        self.totalLength = totalLength
+        self.identification = identification
+        self.flagsAndFragmentOffset = flagsAndFragmentOffset
+        self.timeToLive = timeToLive
+        self.`protocol` = `protocol`
+        self.headerChecksum = headerChecksum
+        self.sourceIpAddress = sourceIpAddress
+        self.destinationIpAddress = destinationIpAddress
+    }
+    
+    init() {
+        self.versionAndIhl = 0
+        self.dscpAndEcn = 0
+        self.totalLength = 0
+        self.identification = 0
+        self.flagsAndFragmentOffset = 0
+        self.timeToLive = 0
+        self.`protocol` = .init(rawValue: 0)
+        self.headerChecksum = 0
+        self.sourceIpAddress = .init(rawValue: 0)
+        self.destinationIpAddress = .init(rawValue: 0)
+    }
+    
+    mutating func setChecksum() {
+        self.headerChecksum = ~[
+            UInt16(versionAndIhl) << 8 | UInt16(dscpAndEcn),
+            totalLength,
+            identification,
+            flagsAndFragmentOffset,
+            UInt16(timeToLive) << 8 | UInt16(`protocol`.rawValue),
+            UInt16(sourceIpAddress.rawValue >> 16),
+            UInt16(sourceIpAddress.rawValue & 0b0000_0000_0000_0000_1111_1111_1111_1111),
+            UInt16(destinationIpAddress.rawValue >> 16),
+            UInt16(destinationIpAddress.rawValue & 0b0000_0000_0000_0000_1111_1111_1111_1111),
+        ].reduce(UInt16(0), onesComplementAdd)
+        
+        assert(isValidChecksum())
+    }
+    
+    func isValidChecksum() -> Bool {
+        let sum = ~[
+            UInt16(versionAndIhl) << 8 | UInt16(dscpAndEcn),
+            totalLength,
+            identification,
+            flagsAndFragmentOffset,
+            UInt16(timeToLive) << 8 | UInt16(`protocol`.rawValue),
+            headerChecksum,
+            UInt16(sourceIpAddress.rawValue >> 16),
+            UInt16(sourceIpAddress.rawValue & 0b0000_0000_0000_0000_1111_1111_1111_1111),
+            UInt16(destinationIpAddress.rawValue >> 16),
+            UInt16(destinationIpAddress.rawValue & 0b0000_0000_0000_0000_1111_1111_1111_1111),
+        ].reduce(UInt16(0), onesComplementAdd)
+        return sum == 0
+    }
+}
+
+extension ByteBuffer {
+    @discardableResult
+    mutating func readIPv4Header() -> IPv4Header? {
+        #if canImport(Darwin)
+        var initialState = self
+        guard
+            let versionAndIhl: UInt8 = self.readInteger(),
```

Comment:
just a question: Why don't you actually use `self.readMultipleIntegers` here?

Reply:
https://www.unix.com/man-page/mojave/4/ip/
> Note that the ip_off and ip_len fields are in host byte order.

relevant comment is at line 189. `self.readMultipleIntegers` sadly doesn't support specifying endianness per integer but I also think it shouldn't as this is a bug in the OS.